CommonTools/ParticleFlow/plugins/PFCandidateRecalibrator.cc

```diff
@@ -0,0 +1,326 @@
+
+	//reset the bad channel containers
+	badChHE_.clear();
+	badChHF_.clear();
+	
+	//fill bad cells HE (use eta, phi)
+	const std::vector<DetId>& cellsHE = hgeom->getValidDetIds(DetId::Detector::Hcal, HcalEndcap);
+	for( auto id : cellsHE)
+	  {
+	    float currentRespCorr = gtCond->getHcalRespCorr(id)->getValue();
+	    float buggedRespCorr = buggedRespCorrs.getValues(id)->getValue();
+	    float ratio = currentRespCorr/buggedRespCorr;
+	    
+	    if(ratio != 1.f)
+	      {
+		GlobalPoint pos = hgeom->getPosition(id);
+		badChHE_.push_back(HEChannel(pos.eta(),pos.phi(),ratio));
+	      }
+	  }
+	
+	//fill bad cells HF (use ieta, iphi)
+	auto const& cellsHF = hgeom->getValidDetIds(DetId::Detector::Hcal, HcalForward);
+	for( auto id : cellsHF)
+	  {
+	    float currentRespCorr = gtCond->getHcalRespCorr(id)->getValue();
+	    float buggedRespCorr = buggedRespCorrs.getValues(id)->getValue();
+	    float ratio = currentRespCorr/buggedRespCorr;
+	    
+	    if(ratio != 1.f)
+	      {
+		HcalDetId dummyId(id);
+		badChHF_.push_back(HFChannel(dummyId.ieta(), dummyId.iphi(), dummyId.depth(), ratio));
+	      }
+	  }
+      }
+}
+
+void PFCandidateRecalibrator::endRun(const edm::Run &iRun, const edm::EventSetup &iSetup)
+{
+}
+
+void PFCandidateRecalibrator::produce(edm::Event &iEvent, const edm::EventSetup &iSetup)
+{
+    //access calogeometry
+    edm::ESHandle<CaloGeometry> calogeom;
+    iSetup.get<CaloGeometryRecord>().get(calogeom);
+    const CaloGeometry* cgeo = calogeom.product();
+    const HcalGeometry* hgeom = static_cast<const HcalGeometry*>(cgeo->getSubdetectorGeometry(DetId::Hcal,HcalForward));
+ 
+    //access PFCandidates
+    edm::Handle<reco::PFCandidateCollection> pfcandidates;
+    iEvent.getByToken(pfcandidates_, pfcandidates);
+
+    int nPfCand = pfcandidates->size();
+    std::unique_ptr<reco::PFCandidateCollection> copy(new reco::PFCandidateCollection());
+    std::unique_ptr<reco::PFCandidateCollection> discarded(new reco::PFCandidateCollection());
+    copy->reserve(nPfCand);
+    std::vector<int> oldToNew(nPfCand), newToOld, badToOld; 
+    newToOld.reserve(nPfCand);
+
+    LogDebug("PFCandidateRecalibrator") << "NEW EV:";
+
+    //loop over PFCandidates
+    int i = -1;
+    for (const reco::PFCandidate &pf : *pfcandidates) 
+      {
+	++i;
+	float absEta = std::abs(pf.eta());
+
+	//deal with HE
+	if( pf.particleId() == reco::PFCandidate::ParticleType::h0 && 
+	    absEta > 1.4  && absEta < 3.)
+	  {
+	    bool toKill = false;
+	    for(auto const& badIt: badChHE_)
+	      if( reco::deltaR2(pf.eta(), pf.phi(), badIt.eta, badIt.phi) < 0.07 )
+		toKill = true;
+	    
+	    
+	    if(toKill)
+	      {
+		discarded->push_back(pf);
+		oldToNew[i] = (-discarded->size());
+		badToOld.push_back(i);
+		continue;
+	      }
+	    else
+	      {
+		copy->push_back(pf);
+		oldToNew[i] = (copy->size());
+		newToOld.push_back(i);
+	      }
+	  }
+	//deal with HF
+	else if(absEta >= 3.)
```

Comment:
from https://github.com/cms-sw/cmssw/pull/22941#discussion_r182119630

> shouldn't this check for particleId 6/7 to be more HF-specific?

Reply:
agreed, will be there at the next commit